README.md
Net-IPAddress-Util
==================

Version-agnostic representation of an IP address in Perl

The goal of the Net::IPAddress::Util modules is to make IP addresses easy to
deal with, regardless of whether they're IPv4 or IPv6, and regardless of the
source (and destination) of the data being manipulated. The module
Net::IPAddress::Util is for working with individual addresses,
Net::IPAddress::Util::Range is for working with individual ranges of
addresses, and Net::IPAddress::Util::Collection is for working with
collections of addresses and/or ranges.

INSTALLATION
============

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

If you would like to run the (time-consuming) battery of time trials, for
instance to decide whether or not it's worthwhile prefering Radix Sorting
over Perl's built-in sort, please set the environment variable
IP_UTIL_TIME_TRIALS to a true value before running 'make test'.

If you do run the time trials, I'd appreciate it if you could email the diag
messages (including the line listing the backend library in use) to me at
paul.w.bennett@gmail.com

SUPPORT AND DOCUMENTATION
=========================

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Net::IPAddress::Util

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Net-IPAddress-Util

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Net-IPAddress-Util

    CPAN Ratings
        http://cpanratings.perl.org/d/Net-IPAddress-Util

    Search CPAN
        http://search.cpan.org/dist/Net-IPAddress-Util/


COPYRIGHT AND LICENCE
=====================

Copyright (C) 2010 Paul Bennett

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

There is _no_ warranty, express or implied.

See <http://dev.perl.org/licenses/> for more information.


Changes
Revision history for Net-IPAddress-Util

5.000   2021-04-05
        v3.x data is not "broken", just "following a different (and more commonly followed) RFC".
        Go back to v3.x being the default, but allow v4.x to be specified with an option-setting method.
        Make the output of Data::Dumper *not* break certain terminals.
        Fix a bug where dot-quad octets beginning with '0' were not being treated as octal.

4.004   2017-10-05
        Enhanced detection and logging of repairable <= v3.x data.

4.003   2017-10-05
        Fix ::Tie so that it actually does what the POD says it does.

4.002   2017-10-04
        Add "repair" functionality to upgrade v3.x data to v4.x format.

4.001   2017-09-22
        Doc fix so that the POD specifies ::ffff:0:0:0/96 to agree with the code.

4.000   2017-09-21
        Major backwards-incompatible bug fix. Was treating ::ffff:0:0/96 as SIIT
        instead of ::ffff:0:0:0/96. Was self-compatible, but RFC-wrong.

3.033   2017-08-31
        Starting to prepare for Perl 5.026

3.032   2017-08-31
        A couple of ultra-paranoid numification tweaks

3.031   2017-08-18
        Fixup whitespace and make the tests smarter

3.030   2017-08-18
        Fix Range->outer_bounds() to return a moderately-compatible object (instead of gibberish), and nudge the tests to reflect that.
        Get rid of weird dangling dependency on utf8 -- we don't use Unicode.

3.029   2017-08-17
        Additional documentation, and a re-check of POD coverage

3.028   2017-08-17
        Clean up versioning in docs & throughout

3.027   2016-04-13
        Clean up versioning in docs & throughout

3.026   2016-04-07
        Respond better to unparseable arguments
        Respond to undef() arguments early, instead of trying to parse them

3.025   2016-03-30
        Trying again to fix stringification of IPv6 addresses

3.024   2016-03-30
        Fix stringification of IPv6 addresses

3.023   2015-03-21
        Remove requirement for Perl 5.12, to see whether it'll work on vanilla 5.10

3.022   2015-03-21
        Solve some extreme edge cases that can arise due to unpacking 'U' and then packing 'C'

3.021   2015-02-18
        Wasn't properly identifying well-formed IPv6 addresses

3.020   2015-01-??
        Code beautifying for Kwalitee purposes

3.019   2015-01-10
        Code beautifying for Kwalitee purposes

3.017   2014-09-29
        Code beautifying for Kwalitee purposes

3.016   2014-09-29
        Code cleaning

3.015   2014-09-29
        Was not accepting integers bigger than 32 bits, regardless of $PROMOTE_N32 setting

3.014   2014-09-27
        Fix typo in documentation: s/to_str/as_str/smg;

        Code cleaning on something that bizarrely appeared to work as desired
        despite being written in a highly nasty way.

3.013   2014-09-27
        Fix RT #99174: IPv6 addresses with fewer than 4 digits per word handled wrong

3.000   2013-11-01
        Back to proper CPAN-friendly version numbers

1.500   2013-10-31
        Attempt to backport the 2.x speedups and fixes into a non-mop env.

1.002   2012-06-09
        Make it explicit that 5.010 is required. Start trying to make the
        time trials smarter.

1.001   2012-05-31
        Try advancing some of the deps to see if that cures the spurious
        test fails on what look like random architectures. This is unrelated
        to the previous GMP / Pari stuff, and might be BigInt version related.

1.000   2012-05-24
        Move from ISA BigInt to HASA BigInt to try to solve the GMP / Pari
        stuff once and for all.

0.12    2011-03-26
        Fix completely idiotic thinko in handling backend specification.

0.11    2011-03-26
        Try to solve weird issue with certain binary versions of GMP and Pari.

0.10    2010-04-28
        Fewer internal dependencies means fewer chances to use the wrong lib.

0.09    2010-03-27
        Try to be even more bullet-proof wrt the N32 bugs.

0.08    2010-03-25
        Add hopefully more-useful output to the diagnostic tests

0.07    2010-03-25
        Add more purely-diagnostic tests.
        Hopefully fix weird $PROMOTE_N32 and ->as_n32() bug on Windows and BSD.

0.06    2010-03-22
        Add $PROMOTE_N32 and ->as_n32().
        Hopefully fix a weird bug on *BSD systems.
        Now at 76.2% coverage.

0.05    2010-03-22
        Make compatability API test better at surviving lack of Net::IPAddress

0.04    2010-03-21
        Added more tests for compatibility API. Now at 75.9% coverage.

0.03    2010-03-21
        Added tests for compatibility API.

0.02    2010-03-21
        Fixed bugs in ip2num() and num2ip(). Fixed distro name in Changes file.

0.01    2010-03-21
        First version, released on an unsuspecting world.


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME                => 'Net::IPAddress::Util',
    AUTHOR              => q{Paul Bennett <paul.w.bennett@gmail.com>},
    VERSION_FROM        => 'lib/Net/IPAddress/Util.pm',
    ABSTRACT_FROM       => 'lib/Net/IPAddress/Util.pm',
    ($ExtUtils::MakeMaker::VERSION >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Carp' => '0',
        'Exporter' => '0',
        'List::MoreUtils' => '0',
        'Math::BigInt' => '0',
        'overload' => '0',
    },
    TEST_REQUIRES => { 'Test::More' => 0, 'Time::HiRes' => '0', },
    MIN_PERL_VERSION => '5.010000',
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'Net-IPAddress-Util-*' },
);


lib/Net/IPAddress/Util/Collection.pm
package Net::IPAddress::Util::Collection;

use strict;
use warnings;
use 5.010;

require Net::IPAddress::Util;
require Net::IPAddress::Util::Collection::Tie;
require Net::IPAddress::Util::Range;

our $VERSION = '5.000';

sub new {
  my $class    = ref($_[0]) ? ref(shift()) : shift;
  my @contents = @_;
  my @o;
  tie @o, 'Net::IPAddress::Util::Collection::Tie', \@contents;
  return bless \@o => $class;
}

sub sorted {
  my $self = shift;
  # In theory, a raw radix sort is O(N), which beats Perl's O(N log N) by
  # a fair margin. However, it _does_ discard duplicates, so ymmv.
  # FIXME Should we sort by hi, lo instead of lo, hi?
  my $from = [ map { [ unpack('C32', $_->{ lower }->{ address } . $_->{ upper }->{ address }) ] } @$self ];
  my $to;
  for (my $i = 31; $i >= 0; $i--) {
    $to = [];
    for my $card (@$from) {
      push @{$to->[ $card->[ $i ] ]}, $card;
    }
    $from = [ map { @{$_ // []} } @$to ];
  }
  my @rv = map {
    my $n = $_;
    my $l = Net::IPAddress::Util->new([@{$n}[0 .. 15]]);
    my $r = Net::IPAddress::Util->new([@{$n}[16 .. 31]]);
    my $x = Net::IPAddress::Util::Range->new({ lower => $l, upper => $r });
    $x;
  } @$from;
  return $self->new(@rv);
}

sub compacted {
  my $self = shift;
  my @sorted = @{$self->sorted()};
  my @compacted;
  my $elem;
  while ($elem = shift @sorted) {
    if (scalar @sorted and $elem->{ upper } >= $sorted[0]->{ lower } - 1) {
      $elem = ref($elem)->new({ lower => $elem->{ lower }, upper => $sorted[0]->{ upper } });
      shift @sorted;
      redo;
    }
    else {
      push @compacted, $elem;
    }
  }
  return $self->new(@compacted);
}

sub tight {
  my $self = shift;
  my @tight;
  map { push @tight, @{$_->tight()} } @{$self->compacted()};
  return $self->new(@tight);
}

sub as_cidrs {
  my $self = shift;
  return map { $_->as_cidr() } grep { eval { $_->{ lower } } } @$self;
}

sub as_netmasks {
  my $self = shift;
  return map { $_->as_netmask() } grep { eval { $_->{ lower } } } @$self;
}

sub as_ranges {
  my $self = shift;
  return map { $_->as_string() } grep { eval { $_->{ lower } } } @$self;
}

1;

__END__

=head1 NAME

Net::IPAddress::Util::Collection - A collection of Net::IPAddress::Util::Range objects

=head1 VERSION

Version 4.004

=head1 SYNOPSIS

  use Net::IPAddress::Util::Collection;

  my $collection = Net::IPAddress::Util::Collection->new();

  while (<>) {
    last if !defined($_);
    push @$collection, $_ if $_;
  }

  print join ', ', $collection->tight->as_ranges;

=head1 DESCRIPTION

Sometimes when dealing with IP Addresses, it can be nice to talk about groups
of them as whole collections of addresses without worrying that the group is
exactly a CIDR-compatible range, or even whether the group is contiguous.

This is what Net::IPAdress::Util::Collection is for. Objects of this class act
as type-checked ARRAYREFs where every entry must be some kind of IP Address
data (either a single address or an arbitrary range).

Knowing that the data within are type-checked (and knowing their specific
type), we can do a few extra things to the ARRAYREF as a whole that we could
not (read I<probably should not>) do to general untyped arrays. Things such as
sorting them via a Radix Sort (which is faster than Perl's builtin sort()),
and being able to smoosh together IP ranges that touch or overlap. What is
more, since we know all IP ranges are ultimately collections of CIDR-compatible
ranges (even if any given range does not start / stop on a legal CIDR boundary)
and use that knowledge to extract precisely the CIDRs that match the collection.

=head1 CLASS METHODS

=head2 new

Create a new Collection object. Takes zero or more arguments, each of which
must be either a L<Net::IPAddress::Util::Range> object, or something which can
be coerced into one (such as a L<Net::IPAddress::Util> object, or something
which can in turn be used to construct one).

=head1 OBJECT METHODS

=head2 sorted

Return a clone of this object, sorted ascendingly by IP address. In the case of
ranges that have the same lower address, ties are broken by the upper address.

=head2 compacted

Return a clone of this object, sorted ascendingly by IP address, with
adjacent ranges combined together. This uses $self-E<gt>sorted, so the same
notice about sort order applies.

=head2 tight

Return a clone of this object, compacted and then split on precise legal CIDR
boundaries. The number of CIDR-compatible ranges returned may be less than,
more than, or in rare cases the same as the number of elements in the original
Collection object. Such is the CIDR nature.

=head2 as_ranges

Returns an array of stringified (x .. y) style ranges.

=head2 as_cidrs

Returns an array of stringified CIDR-style strings. In the case where one
element of the Collection cannot be legally represented as a CIDR, you will get
in its place the smallest single legal CIDR that contains that element.

In other words, if you want complete accuracy, you will want to use:

  $collection->tight->as_cidrs;

=head2 as_netmasks

Returns an array of stringified Netmask-style strings. In the case where one
element of the Collection cannot be legally represented as a Netmask string,
you will get in its place the smallest single legal Netmask string that
contains that element.


In other words, if you want complete accuracy, you will want to use:

  $collection->tight->as_netmasks;

=cut


lib/Net/IPAddress/Util/Collection/Tie.pm
package Net::IPAddress::Util::Collection::Tie;

use strict;
use warnings;
use 5.012;
use utf8;

use Carp qw( confess );

sub new {
    my $class = shift;
    $class = ref($class) || $class;
    my ($arg_ref) = @_;
    return bless $arg_ref => $class;
}

sub TIEARRAY {
    my ($class, $contents) = @_;
    $contents = [] unless defined $contents;
    @{$contents} = map { _checktype($_) } @{$contents};
    my $self = $class->new({ contents => $contents });
}

sub FETCH {
    my ($self, $i) = @_;
    return $self->{ contents }->[ $i ];
}

sub STORE {
    my ($self, $i, $v) = @_;
    $self->{ contents }->[ $i ] = _checktype($v);
    return $v;
}

sub FETCHSIZE {
    my ($self) = @_;
    return scalar @{$self->{ contents }};
}

sub EXISTS {
    my ($self, $i) = @_;
    return exists $self->{ contents }->[ $i ];
}

sub DELETE {
    my ($self, $i) = @_;
    return delete $self->{ contents }->[ $i ];
}

sub CLEAR {
    my ($self) = @_;
    $self->{ contents } = [ ];
    return $self->{ contents };
}

sub PUSH {
    my ($self, @l) = @_;
    push @{$self->{ contents }}, map { _checktype($_) } @l;
}

sub POP {
    my ($self) = @_;
    return pop @{$self->{ contents }};
}

sub UNSHIFT {
    my ($self, @l) = @_;
    unshift @{$self->{ contents }}, map { _checktype($_) } @l;
}

sub SHIFT {
    my ($self) = @_;
    return shift @{$self->{ contents }};
}

sub SPLICE {
    my ($self, $offset, $length, @l) = @_;
    $offset = 0 unless defined $offset;
    $length = $self->FETCHSIZE() - $offset unless defined $length;
    return splice @{$self->{ contents }}, $offset, $length, map { _checktype($_) } @l;
}

sub _checktype {
    my ($v) = @_;
    return $v if ref $v eq 'Net::IPAddress::Util::Range';
    if (ref $v eq 'HASH') {
        $v = Net::IPAddress::Util::Range->new($v);
    }
    if (ref $v eq 'Net::IPAddress::Util') {
        $v = Net::IPAddress::Util::Range->new({ ip => $v });
    }
    if (!defined $v or ref $v ne 'Net::IPAddress::Util::Range') {
        my $disp = defined $v ? ref $v ? ref $v : 'bare scalar' : 'undef()';
        confess("Invalid data type ($disp)");
    }
    return $v;
}

1;

__END__

=head1 NAME

Net::IPAddress::Util::Collection::Tie - These aren't the droids you're looking for

=head1 METHODS

=head2 new

No, seriously. You should not be poking around back here. You are likely to be eaten by a grue.

=cut



lib/Net/IPAddress/Util/Range.pm
package Net::IPAddress::Util::Range;

use strict;
use warnings;
use 5.010;

use overload (
  '""' => 'as_string',
  '<=>' => '_spaceship',
  'cmp' => '_spaceship',
);

use Net::IPAddress::Util qw( :constr :manip );
require Net::IPAddress::Util::Collection;

our $VERSION = '5.000';

sub new {
  my $class = shift;
  $class = ref($class) || $class;
  my ($arg_ref) = @_;
  my ($l, $u);
  if ($arg_ref->{ lower } && $arg_ref->{ upper }) {
    $arg_ref->{ lower } = IP($arg_ref->{ lower });
    $arg_ref->{ upper } = IP($arg_ref->{ upper });
    if ($arg_ref->{ lower } > $arg_ref->{ upper }) {
      ($arg_ref->{ lower }, $arg_ref->{ upper }) = ($arg_ref->{ upper }, $arg_ref->{ lower });
    }
    return bless $arg_ref => $class;
  }
  elsif ($arg_ref->{ ip }) {
    my $ip;
    my $nm = 2;
    if ($arg_ref->{ netmask }) {
      $ip = IP($arg_ref->{ ip      });
      my $was_ipv4 = $ip->is_ipv4;
      $nm = IP($arg_ref->{ netmask });
      $ip &= $nm;
      $nm = ~$nm;
      if ($was_ipv4) {
        $nm &= ipv4_mask();
      }
      $l = $ip;
      $u = $ip | $nm;
    }
    elsif ($arg_ref->{ ip } =~ m{(.*?)/(\d+)}) {
      my ($t, $cidr) = ($1, $2);
      $ip = IP($t);
      my $was_ipv4 = $ip->is_ipv4;
      my $span
        = ($was_ipv4
        ? 32
        : 128) - $cidr
        ;
      $nm = implode_ip(substr(('1' x 128) . ('0' x $span), -128));
      $ip &= $nm;
      $l = $ip;
      $u = $ip | ~$nm;
    }
    elsif ($arg_ref->{ cidr }) {
      $ip = IP($arg_ref->{ ip });
      my $was_ipv4 = $ip->is_ipv4;
      my $cidr = $arg_ref->{ cidr };
      my $span
        = ($was_ipv4
        ? 32
        : 128) - $cidr
        ;
      $nm = implode_ip(substr(('1' x 128) . ('0' x $span), -128));
      $ip &= $nm;
      $l = $ip;
      $u = $ip | ~$nm;
    }
    else {
      $l = IP($arg_ref->{ ip });
      $u = IP($arg_ref->{ ip });
    }
  }
  return bless { lower => $l, upper => $u } => $class;
}

sub as_string {
  my $self = shift;
  return "($self->{ lower } .. $self->{ upper })";
}

sub outer_bounds {
  my $self = shift;
  my @l = explode_ip($self->{ lower });
  my @u = explode_ip($self->{ upper });
  my @cidr = common_prefix(@l, @u);
  my $cidr = scalar @cidr;
  my $base = implode_ip(ip_pad_prefix(@cidr));
  if ($base->is_ipv4()) {
    $cidr -= 96;
  }
  my @mask = prefix_mask(@l, @u);
  my $nm = implode_ip(ip_pad_prefix(@mask));
  my $x = ~$nm;
  if ($base->is_ipv4()) {
    $nm &= ipv4_mask();
  }
  my $hi = IP($base);
  $hi |= $x;
  return bless {
    lower   => $base,
    cidr    => $cidr,
    netmask => $nm,
    upper   => $hi,
  } => ref($self);
}

sub inner_bounds {
  my $self = shift;
  return $self if $self->{ upper } == $self->{ lower };
  my $bounds = $self->outer_bounds();
  my $new = ref($self)->new($self);
  while ($bounds->{ upper } > $self->{ upper } or $bounds->{ lower } < $self->{ lower }) {
    $new = ref($self)->new({ ip => $self->{ lower }, cidr => $bounds->{ cidr } + 1 });
    $bounds = $new->outer_bounds();
  }
  return $new;
}

sub as_cidr {
  my $self = shift;
  my $hr = $self->outer_bounds();
  return "$hr->{ lower }" . '/' . "$hr->{ cidr }";
}

sub as_netmask {
  my $self = shift;
  my $hr = $self->outer_bounds();
  return "$hr->{ lower }" . ' (' . "$hr->{ netmask }" . ')';
}

sub loose {
  my $self = shift;
  my $hr = $self->outer_bounds();
  return ref($self)->new({ lower => $hr->{ lower }, upper => $hr->{ upper } });
}

sub _spaceship {
  my ($self, $rhs, $swapped) = @_;
  ($self, $rhs) = ($rhs, $self) if $swapped;
  $rhs = ref($self)->new({ ip => $rhs }) unless ref($self) eq ref($rhs);
  return
    $self->{ lower } <=> $rhs->{ lower }
    || $self->{ upper } <=> $rhs->{ upper }
    ;
}

sub tight {
  my $self = shift;
  my $inner = $self->inner_bounds();
  my $rv = Net::IPAddress::Util::Collection->new();
  push @$rv, $inner;
  if ($inner->{ upper } < $self->{ upper }) {
    my $remainder = ref($self)->new({ lower => $inner->{ upper } + 1, upper => $self->{ upper } });
    push @$rv, @{$remainder->tight()};
  }
  return $rv;
}

sub lower {
  my $self = shift;
  return $self->{ lower };
}

sub upper {
  my $self = shift;
  return $self->{ upper };
}

1;

__END__

=head1 NAME

Net::IPAddress::Util::Range - Representation of a range of IP addresses

=head1 VERSION

Version 4.004

=head1 SYNOPSIS

  use Net::IPAddress::Util::Range;

  my $x = '192.168.0.3';
  my $y = '192.168.0.123';

  my $range = Net::IPAddress::Util::Range->new({ lower => $x, upper => $y });

  print "$range\n"; # (192.168.0.3 .. 192.168.0.123)

  for (@{$range->tight()}) {
    print "$_\n";
  }

  my $w = '192.168.0.0/24';

  my $range = Net::IPAddress::Util::Range->new({ ip => $w });

  my $v = '192.168.0.0';

  my $range = Net::IPAddress::Util::Range->new({ ip => $v, cidr => 24 });

  my $z = '255.255.255.0';

  my $range = Net::IPAddress::Util::Range->new({ ip => $v, netmask => $z });

=head1 DESCRIPTION

Sometimes when dealing with IP Addresses, it can be nice to talk about
contiguous ranges of them as whole collections of addresses without worrying
that the contiguous range is exactly a CIDR-compatible range.

This is what Net::IPAdress::Util::Range is for. Objects of this class act
as type-checked pairs of lower and upper bounds of a range of IP Addresses.

=head1 CLASS METHODS

=head2 new

The constructor. Takes a hashref with:

=over

=item C<lower> and C<upper>

In this case, construction is straightforward. The two values must be either
L<Net::IPAddress::Util> objects, or something that can be used to construct one.

=item C<ip>

If the C<ip> is a L<Net::IPAddress::Util> object (or something that can be used
to construct one), then you get a Range consisting of a single IP. This may
seem redundant, but allows L<Net::IPAddress::Util::Collection> to do magic.

=over

Also, as a convenience, CIDR strings (of the form "N.N.N.N/X", etc) may be
passed in, and you'll get back a Range representing that whole CIDR.

=back

=item C<ip> and C<cidr>

Pass in an IP and a numeric CIDR (which B<MUST> be valid for the version (4 or
6) of the IP), and you'll get back a Range representing that whole CIDR.

=item C<ip> and C<netmask>

Pass in two IPs, of the same version, and they'll be treated exacatly as the
argument names suggest. The C<netmask> argument B<MUST> (in binary) start with
zero or more ones, followed by enough zeroes to pad out to the correct number
of bits (either 32 or 128 for IPv4 or IPv6 respectively). The C<ip> argument
B<MUST> have the same number of right-hand zeroes as the C<netmask> argument.

=back

=head1 OBJECT METHODS

=head2 '""'

=head2 as_string

Objects stringify to a representation of their range.

=head2 as_cidr

Stringification for CIDR-style strings.

=head2 as_netmask

Stringification for Netmask-style strings.

=head2 outer_bounds

Return the bounds of the smallest subnet capable of completely containing
the addresses in this range. Note that this is not automatically the same
thing as "the subnet that matches this range", as a range may or may not be
aligned to legal subnet boundaries.

=head2 inner_bounds

Return the bounds of the largest subnet capable of being completely contained
by the addresses in this range. Note that this is not automatically the same
thing as "the subnet that matches this range", as a range may or may not be
aligned to legal subnet boundaries.

=head2 tight

Returns a collection of subnets that (between them) exactly match the
addresses in this range. The returned object is a L<Net::IPAddress::Util::Collection>,
which can be treated as an array reference of Net::IPAddress::Util::Range objects.

=head2 loose

Returns a blessed object (of this class) representing the range returned by outer_bounds().

=head2 lower

=head2 upper

Get the lower or upper bounds of this range.

=cut
